derivers.lua

```diff
@@ -143,8 +237,187 @@ local is = {
 			assert(compiled, message)
 			idx["is_" .. vname] = compiled()
 		end
+
+		t.derived_is = true
+	end,
+	foreign = function()
+		error("can't derive is for a foreign type")
+	end,
+	map = function()
+		error("can't derive is for a map type")
+	end,
+	set = function()
+		error("can't derive is for a set type")
+	end,
+	array = function()
+		error("can't derive is for an array type")
+	end,
+}
+
+---@type Deriver
+local unwrap = {
+	record = function(t, info)
+		if t.derived_unwrap then
+			-- already derived, don't derive again
+			return
+		end
+
+		local idx = t.__index or {}
+		t.__index = idx
+		local kind = info.kind
+		local params = info.params
+
+		local returns = {}
+		for i, param in ipairs(params) do
+			returns[i] = string.format("self[%q]", param)
+		end
+		local all_returns = table.concat(returns, ", ")
+
+		local chunk = string.format("return function(self) return %s end", all_returns)
+
+		derive_print("derive unwrap: record chunk: " .. kind)
+		derive_print("###")
+		derive_print(chunk)
+		derive_print("###")
+
+		local compiled, message = load(chunk, "derive-unwrap_record", "t")
+		assert(compiled, message)
+		idx["unwrap_" .. kind] = compiled()
+		t.derived_unwrap = true
+	end,
+	enum = function(t, info)
+		if t.derived_unwrap then
+			-- already derived, don't derive again
+			return
+		end
+
+		local idx = t.__index or {}
+		t.__index = idx
+		local name = info.name
+		local variants = info.variants
+
+		for _, vname in ipairs(variants) do
+			local vkind = name .. "." .. vname
+			local vdata = variants[vname]
+			local vtype = vdata.type
+			local vinfo = vdata.info
+			local all_returns
+			if vtype == EnumDeriveInfoVariantKind.Record then
+				local returns = {}
+				for i, param in ipairs(vinfo.params) do
+					returns[i] = string.format("self[%q]", param)
+				end
+				all_returns = table.concat(returns, ", ")
+			elseif vtype == EnumDeriveInfoVariantKind.Unit then
+				all_returns = ""
+			else
+				error("unknown variant type: " .. vtype)
+			end
+
+			local chunk = [[
+return function(self)
+	if self.kind == %q then
+		return %s
+	else
+		error("unwrap failed: unwrapping for a %s but found a " .. self.kind)
+	end
+end]]
+			chunk = chunk:format(vkind, all_returns, vkind)
+
+			derive_print("derive unwrap: enum chunk: " .. vkind)
+			derive_print("###")
+			derive_print(chunk)
+			derive_print("###")
+
+			local compiled, message = load(chunk, "derive-unwrap_enum", "t")
+			assert(compiled, message)
+			idx["unwrap_" .. vname] = compiled()
+		end
+
+		t.derived_unwrap = true
+	end,
+	foreign = function()
+		error("can't derive unwrap for a foreign type")
```

Comment:
change to "can't derive unwrap()" for more clarity

Reply:
resolved in #95